### PR TITLE
Nerfs Juggling

### PR DESCRIPTION
**Pull request by RikuTheKiller**
## About The Pull Request

Makes it so that firing a different shotgun before the firing delay of the shotgun you last fired has passed now causes 20 brute damage to the arm you're using alongside a warning message in chat. This ONLY affects juggling.

The message reads "Your arm is torn by the recoil!" in big, red text.

You can still get around 3 - 4 shots before a fracture and you'll either go critical or delimb if you go another 3 - 4 shots. This means that if you somehow manage to accidentally juggle a shot or two you won't just, idk, fucking explode?

This does take into account akimbo and has a check that prevents akimbo from being affected, don't worry about that.

Even if the limb has taken enough burn damage to avoid delimbing through ordinary means, this will forcefully delimb if it's at max damage when dealing damage. This prevents circumventing the punishment by using burn damage.

## Why It's Good For The Game

This should be obvious. If a solo marine can kill a gorger with no issues in less than a second using juggling it should be pretty obvious why nerfing it/removing it is a good idea. In terms of balancing, 1 xeno is meant to be 3 marines and we're talking about the tankiest T3 xeno, so this just doesn't make any fucking sense and is annoying as all hell. (Primo crushers can also be reliably solo'd by this.)

So many people have complained about this that it's unreal. This is the first PR that tries to fix this issue that DOESN'T affect other aspects of the gun, such as having to wield it before firing. Things such as that significantly affect the weapon's CQC potential and may ruin it's viability while only partially mitigating the actual issue.

## Changelog
:cl:
balance: Juggling now has a 20 damage penalty to the arm you're using every time you do it.
/:cl:

### CODE REVIEW

**Comment by Losenis** [comment]
If you can take the firerate of one shotgun and have it check for the next, wouldn't it be both safer and more intuitive to just force the firerate of the last gun to also affect the shot of the next gun? Would also have some forward compatibility with funny things like juggling martinis or mosins or whatever other non-shotgun gun.

**Review by Pariah919 — CHANGES_REQUESTED** [review]
Denying this on the sole fact that in balance gotcha mechanics are stupid and bad

**Comment by FIXTHERADBOXPLEASE** [comment]
Losenis that doesn't work with DB juggling. The DB has a rapid firerate that's slowed solely by the fact it has two shots before reloading. It'd still have the problem even with enforced firerates.

Not a coder but while I'd want to agree with Pariah, I can't think of a better solution. Unless maybe the DB's actual firerate is nerfed, but that'd be stupid IMO.

Maybe nerf it to 5, 10 damage or something. 20 is alot, and doing any amount of self-damage during combat is liable to get you punished hard by the xenos.

Maybe have it so any shots during the firing delay just have extreme inaccuracy that makes them ALWAYS miss, instead of doing self-damage. This makes the ""mechanic"" non-viable, while not particularly punishing the marine for doing it. It'd still work with PBs but juggling is mostly done with slugs anyway.

**Comment by lovegreenstuff** [comment]
aight so not to be that guy, but like...

if you have a variable for last fired that isnt tied to a gun, why not just make it so that you can't fire it at all

**Comment by FIXTHERADBOXPLEASE** [comment]
> if you have a variable for last fired that isnt tied to a gun, why not just make it so that you can't fire it at all

Because that only fixes juggling for combat and pump shotguns. It barely affects juggling DBs.
The DB has a rapid firerate, kept in check only by its low ammo. Juggling DBs allows you to fire it up to six times in a row. With that change, shooting at the DB's fire delay, which is still insanely fast.

Although I'm stupid because on second thought even this wouldn't fix DB juggling, since the jugglers could just shoot one DB twice before swapping to the next. Every shot would still be at the DB's fast firerate.

**Comment by Al-1ce** [comment]
This seems to be the only fix to this problem that doesn't adjust/fuck up other things that have nothing to do with the problem.

People do not want the shotguns to be nerfed. They want juggling to be removed. This PR does that without hurting or affecting any other variables _and yet people still complain._

**Comment by RikuTheKiller** [comment]
> This looks to be the only fix for this problem that doesn't adjust/fuck up other things that have nothing to do with the actual issue..
> 
> People do not want the shotguns to be nerfed. They want juggling to be removed. This PR does that without hurting or affecting any other variables _and YET people still complain._

But apparently when someone made DBs only able to shoot when wielded it was fine even though people had a fucking cope seizure over the fact that DB was no longer viable?

Oh, also, this PR isn't meant to fix juggling things for ammo efficiency. If you want to waste an entire gun slot for a few more ammo, sure, that's a rather big thing.

This PR keeps juggling as a LAST RESORT mostly for people who dual wield. Since a marine's life is worth a lot less than a xeno's it may still be worth it to juggle in some cases, however, you'll fuck yourself over.

I CAN stop people from firing, by the way, but that's just not a thing I want to do. Many people said this would be closed since the head coder considers juggling a feature. So I simply wanted to nerf it. If anyone wants to go even further and suggest a full removal, go right ahead. That's out of scope for this PR. Reverting this PR in that PR wouldn't even be hard. You just remove the defines and replace the method override and you're good!

**Comment by RikuTheKiller** [comment]
> Maybe nerf it to 5, 10 damage or something. 20 is alot, and doing any amount of self-damage during combat is liable to get you punished hard by the xenos.
> 
> Maybe have it so any shots during the firing delay just have extreme inaccuracy that makes them ALWAYS miss, instead of doing self-damage. This makes the ""mechanic"" non-viable, while not particularly punishing the marine for doing it. It'd still work with PBs but juggling is mostly done with slugs anyway.

5 or 10 damage isn't enough to justify a stunlock. At that point your damage output is like 5 - 8 times what you're getting punished for. The fact I even let it be a net positive should be good enough. (You can still juggle with a broken arm if you REALLY want to.)

**Comment by RikuTheKiller** [comment]
By the way, please stop trying to throw random alternate methods into the air. It simply won't work. It has been tried like 2 - 3 times before and every single time it's either had repercussions or been closed. Just think about what you say in your head before typing it out.

**Comment by Lumipharon** [comment]
As hilarious as this sounds, and being probably a bad fix, it doesn't actually work at all.
Limbs have a damage cap, and normally if they get enough brute they break and eventually fall off.

However this can be bypassed. You can apply some burn to your arm so you can't reach that threshold, have a robotic arm or simply be a combat robot.

Dropping shit in your hand from a frac/malfunction doesn't matter since you're only ever holding an item for 0.1 seconds anyway.

**Comment by RikuTheKiller** [comment]
> As hilarious as this sounds, and being probably a bad fix, it doesn't actually work at all. Limbs have a damage cap, and normally if they get enough brute they break and eventually fall off.
> 
> However this can be bypassed. You can apply some burn to your arm so you can't reach that threshold, have a robotic arm or simply be a combat robot.
> 
> Dropping shit in your hand from a frac/malfunction doesn't matter since you're only ever holding an item for 0.1 seconds anyway.

Did you not like... read any context? The fracturing is intentional AND if you keep doing it then you'll either delimb or go into critical. "if they get enough brute they break and eventually fall off" that's the point. Read the PR description. If you keep doing it even after the warning damage you'll get a fracture and then a delimb or critical.

**Comment by RikuTheKiller** [comment]
Alright it now forcefully delimbs if the limb is at max damage when doing the punishment. This prevents the burn damage issue.

**Comment by RikuTheKiller** [comment]
> Right now every single complain I read here is literally a whining from someone that probably personally abuses juggling himself.

Pretty much, yeah.

**Comment by UltraBigRat** [comment]
Good pr

**Comment by Naaanii** [comment]
Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.

**Comment by RikuTheKiller** [comment]
> Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.

Woo lets overall nerf a gun that doesn't have any problems in order to solve a different problem! I'm great at game design! It's almost like the last time someone did the exact nerf you just submitted everyone complained!

Honestly, I am so done with that bullshit, fuck you.

**Comment by RikuTheKiller** [comment]
> By the way, please stop trying to throw random alternate methods into the air. It simply won't work. It has been tried like 2 - 3 times before and every single time it's either had repercussions, never even been tried since the idea was just dead beforehand or been made and then closed. Just think about what you say in your head before typing it out.

Just to add salt to the wound I'll quote a reply from this exact comment section.

**Comment by RikuTheKiller** [comment]
> As hilarious as this sounds, and being probably a bad fix, it doesn't actually work at all. Limbs have a damage cap, and normally if they get enough brute they break and eventually fall off.
> 
> However this can be bypassed. You can apply some burn to your arm so you can't reach that threshold, have a robotic arm or simply be a combat robot.
> 
> Dropping shit in your hand from a frac/malfunction doesn't matter since you're only ever holding an item for 0.1 seconds anyway.

All issues from this comment have been addressed, none of them apply. Fractures don't matter? Then crit or a delimb sure will. It also bypasses burn damage now.

"probably a bad fix" how about you try it in-game or something then.

**Comment by RikuTheKiller** [comment]
Also, I'd like for this PR to be test merged to see how it would actually function in-game. (Since this is pretty major.)

**Comment by Weedmaster77** [comment]
I like the idea of limb damage for doing it but I don't like the delimbing aspect seems like a bit much maybe play more to paincrit instead of delimb if that can be done or even add IB to the arm along with it getting broken but delimbing is just a pain for the team as a whole and medics as xenos do it enough we don't need self delimbing IMO

**Comment by Naaanii** [comment]
> > Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.
> 
> Woo lets overall nerf a gun that doesn't have any problems doing it's job (CQC) in order to solve a different problem! I'm great at game design! It's almost like the last time someone did the exact nerf you just submitted everyone complained!
> 
> This exact thing has been discussed so many times. Yet it always comes to the same conclusion, it nerfs shotguns overall.
> 
> Honestly, I am so done with that bullshit, fuck you.
 
It doesn't nerf shotguns overall and nerfs single handing shotguns very slightly. Nerfing ammo types would nerf shotguns overall. 

MJP already gave the go ahead to remove one handedness from the combat shotty.

**Comment by RikuTheKiller** [comment]
> I like the idea of limb damage for doing it but I don't like the delimbing aspect seems like a bit much maybe play more to paincrit instead of delimb if that can be done or even add IB to the arm along with it getting broken but delimbing is just a pain for the team as a whole and medics as xenos do it enough we don't need self delimbing IMO

The delimbing only happens if you couldn't give less of a shit about the damage. You get to juggle 7 times before it delimbs. (8 if you count the first shot as juggling.)

**Comment by RikuTheKiller** [comment]
> > > Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.
> > 
> > 
> > Woo lets overall nerf a gun that doesn't have any problems doing it's job (CQC) in order to solve a different problem! I'm great at game design! It's almost like the last time someone did the exact nerf you just submitted everyone complained!
> > This exact thing has been discussed so many times. Yet it always comes to the same conclusion, it nerfs shotguns overall.
> > Honestly, I am so done with that bullshit, fuck you.
> 
> It doesn't nerf shotguns overall and nerfs single handing shotguns very slightly. Nerfing ammo types would nerf shotguns overall.
> 
> MJP already gave the go ahead to remove one handedness from the combat shotty.

"Nerfs single handing shotguns very slightly" no it REMOVES single handing shotguns.

**Comment by RikuTheKiller** [comment]
> > > > Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.
> > > 
> > > 
> > > Woo lets overall nerf a gun that doesn't have any problems doing it's job (CQC) in order to solve a different problem! I'm great at game design! It's almost like the last time someone did the exact nerf you just submitted everyone complained!
> > > This exact thing has been discussed so many times. Yet it always comes to the same conclusion, it nerfs shotguns overall.
> > > Honestly, I am so done with that bullshit, fuck you.
> > 
> > 
> > It doesn't nerf shotguns overall and nerfs single handing shotguns very slightly. Nerfing ammo types would nerf shotguns overall.
> > MJP already gave the go ahead to remove one handedness from the combat shotty.
> 
> "Nerfs single handing shotguns very slightly" no it REMOVES single handing shotguns.

Not to mention removing shotgun + shield combo.

**Comment by Naaanii** [comment]
> > > > Just make it so you can't fire mid wield and give issue guns higher wield time. Combat shotty can have it's one handedness removed.
> > > 
> > > 
> > > Woo lets overall nerf a gun that doesn't have any problems doing it's job (CQC) in order to solve a different problem! I'm great at game design! It's almost like the last time someone did the exact nerf you just submitted everyone complained!
> > > This exact thing has been discussed so many times. Yet it always comes to the same conclusion, it nerfs shotguns overall.
> > > Honestly, I am so done with that bullshit, fuck you.
> > 
> > 
> > It doesn't nerf shotguns overall and nerfs single handing shotguns very slightly. Nerfing ammo types would nerf shotguns overall.
> > MJP already gave the go ahead to remove one handedness from the combat shotty.
> 
> "Nerfs single handing shotguns very slightly" no it REMOVES single handing shotguns.

Yeah I meant 2 handing.

Shields and shotguns shouldn't have existed to begin with and just attributed to akimbo and juggling issues. Either way, Green's idea was better, since you have the code and have tested it, to just stop the ability of firing shotguns in quick succession.

**Comment by TiviPlus** [comment]
Given the low code quality and shitposted effect I am filing this away as a shitpost

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 1]
```diff
@@ -128,3 +128,10 @@
 
 	/// This is the cooldown on suffering additional effects for when shock gets high
 	COOLDOWN_DECLARE(last_shock_effect)
+
+///This shows when the human last fired a shotgun.
+	var/lastshotguntime
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/human/human_defines.dm:133]: Shouldnt affect only shotguns

File: code/modules/projectiles/guns/shotguns.dm [hunk 2]
```diff
@@ -33,6 +33,20 @@
...
 
 
+/obj/item/weapon/gun/shotgun/do_fire() //Handles recoil punishments for juggling.
+	. = ..()
+	var/mob/living/carbon/human/wielder = gun_user
+	if (wielder.lastshotgun != src && wielder.lastshotgun != wielder.get_inactive_held_item() && world.time < wielder.lastshotguntime + wielder.lastshotgundelay) //Does punishment for juggling in the form of recoil damage.
+		var/zone = wielder.hand ? "l_arm" : "r_arm"
+		var/datum/limb/limb = wielder.get_limb(zone)
+		to_chat(wielder, span_userdanger("Your arm is torn by the recoil!"))
+		wielder.apply_damage(20, BRUTE, zone)
```
Comment (BraveMole) [inline code/modules/projectiles/guns/shotguns.dm:43]: Should not make damage, certainly not. That is a gotcha mecanic and thats a no.

Just prevent from firing

File: code/modules/projectiles/guns/shotguns.dm [hunk 3]
```diff
@@ -33,6 +33,20 @@
 	placed_overlay_iconstate = "shotgun"
 
 
+/obj/item/weapon/gun/shotgun/do_fire() //Handles recoil punishments for juggling.
```
Comment (BraveMole) [inline code/modules/projectiles/guns/shotguns.dm:36]: Shouldnt be on shotgun

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by Alexeivichi** [comment]
Oh boy, this is one hell of a PR.

I dislike it because of it's extreme punishment (20 brute is insane, there's more ways to just remedy the juggling meta) but I appreciate the attempt to fix it overall.

It'd generally be better to cause the marine to get stunned for a brief period of time if they attempt to fire 2 shotguns like this. You could just probably remove the 1-handedness of the T-39 shotgun. Maybe apply a heavier recoil to the screen? If you're able to track the firing of shotguns one after the other for juggling, other effects could work? Even just applying a high inaccuracy bonus upon wielding a shotgun for the first second (or two if you think it's that bad) could stop this.

Regardless, causing 20 brute damage is way too much. A literal aggressive slash with a knife or bullet to my arm does less damage than that as I'm wearing body armour.

I'd still say the best way to punish juggling is heavy recoil stacking, or applying a soft stun for a short time.

File: code/modules/mob/living/carbon/human/human_defines.dm (hunk 1 above)
Comment (lKiseki) [reply code/modules/mob/living/carbon/human/human_defines.dm:133]: > Shouldnt affect only shotguns

Regarding this; marines have been juggling literally any weapon they can get their hands on. TL-127s, Grenade Launchers, and so on. It'd be a good idea to have this PR encompass all guns so we can end the issue altogether.

**Comment by ivanmixo** [comment]
Both the code and the idea suck

**Comment by RikuTheKiller** [comment]
I'll be changing this to another removal PR since I think I've got a pretty good idea.

**Comment by RikuTheKiller** [comment]
Actually, nah.

**Pull request closed without merging**